src/chapters/adding-protocol.xml

```diff
@@ -118,385 +221,564 @@ AM_CFLAGS = -Werror -Wall -pedantic
+            maintains the state of the ball, we must create a header file,
+                <filename>src/ball.h</filename>, to define this:</para>
         <informalexample>
-        <programlisting xml:id="ball-03-ball_client.c" version="5.0" xml:lang="en">int guac_client_init(guac_client* client, int argc, char** argv) {
-<emphasis>
-    /* The layer which will contain our ball */
+            <programlisting version="5.0" xml:lang="en">#ifndef BALL_H
+#define BALL_H
+
+#include &lt;guacamole/layer.h>
+
+typedef struct ball_client_data {
+
     guac_layer* ball;
+
+} ball_client_data;
+
+#endif</programlisting>
+        </informalexample>
+        <para>To make the build system aware of the existence of the new
+                <filename>src/ball.h</filename> header file, <filename>Makefile.am</filename> must
+            be updated as well:</para>
+        <informalexample>
+            <programlisting version="5.0" xml:lang="en">...
+
+# All source files of libguac-client-ball
+<emphasis>noinst_HEADERS = src/ball.h</emphasis>
+libguac_client_ball_la_SOURCES = src/ball.c
+
+...</programlisting>
+        </informalexample>
+        <para>This new structure is intended to house the client-level state of the ball,
+            independent of any users which join or leave the connection. The structure must be
+            allocated when the client begins (within <function>guac_client_init</function>), freed
+            when the client terminates (via a new client free handler), and must contain the layer
+            which represents the ball within the remote display. As this layer is part of the remote
+            display state, it must additionally be initialized when a user joins, in the same way
+            that the display overall was initialized in earlier steps:</para>
+        <informalexample>
+        <programlisting xml:id="ball-03-ball_client.c" version="5.0" xml:lang="en"><emphasis>#include "ball.h"</emphasis>
+
+#include &lt;guacamole/client.h>
+<emphasis>#include &lt;guacamole/layer.h></emphasis>
+#include &lt;guacamole/protocol.h>
+#include &lt;guacamole/socket.h>
+#include &lt;guacamole/user.h>
+
+#include &lt;stdlib.h>
+
+...
+
+int ball_join_handler(guac_user* user, int argc, char** argv) {
+
+    /* Get client associated with user */
+    guac_client* client = user->client;
+<emphasis>
+    /* Get ball layer from client data */
+    ball_client_data* data = (ball_client_data*) client->data;
+    guac_layer* ball = data->ball;
 </emphasis>
     ...
 <emphasis>
-    /* Set up our ball layer */
-    ball = guac_client_alloc_layer(client);
-    guac_protocol_send_size(client->socket, ball, 128, 128);
+    /* Set up ball layer */
+    guac_protocol_send_size(socket, ball, 128, 128);
 
     /* Fill with solid color */
-    guac_protocol_send_rect(client->socket, ball,
+    guac_protocol_send_rect(socket, ball,
             0, 0, 128, 128);
 
-    guac_protocol_send_cfill(client->socket,
+    guac_protocol_send_cfill(socket,
```

Comment:
Isn't this the instruction that fills the color?

And the other one creates a rectangle?

Comments are a little confusing.